src/uffd_handler/src/handler.rs

```diff
@@ -0,0 +1,432 @@
+    EventNotReady,
+    #[error("Failed to fetch system's page size: {0}")]
+    PageSize(utils::errno::Error),
+    #[error("Parsing data stream failed: {0}")]
+    ParseStream(StreamError),
+    #[error("Failed to obtain Firecracker's credentials.")]
+    PeerCredentials,
+    #[error("Failed to read userfaultfd event: {0}.")]
+    ReadEvent(userfaultfd::Error),
+    #[error("Userfaultfd copy failed: {0}")]
+    UffdCopy(userfaultfd::Error),
+    #[error("Userfaultfd {0} operation returned 0 bytes.")]
+    UffdNoOperation(String),
+    #[error("Userfaultfd failed to zero out pages: {0}")]
+    UffdZero(userfaultfd::Error),
+    #[error("Userfaultfd received unexpected event: {0:?}.")]
+    UnexpectedEvent(Event),
+}
+
+type Result<T> = std::result::Result<T, HandlerError>;
+
+pub trait UffdManager: AsRawFd {
+    fn poll_fd(&self) -> Result<Event>;
+
+    fn populate_from_file(&self, buff: u64, region: &MemRegion) -> Result<(u64, u64)>;
+
+    fn zero_out(&self, addr: u64) -> Result<(u64, u64)>;
+}
+
+impl UffdManager for Uffd {
+    fn poll_fd(&self) -> Result<Event> {
+        // Read an event from the userfaultfd.
+        let event_op = self.read_event().map_err(HandlerError::ReadEvent)?;
+        match event_op {
+            Some(event) => Ok(event),
+            None => Err(HandlerError::EventNotReady),
+        }
+    }
+
+    fn populate_from_file(&self, buff: u64, region: &MemRegion) -> Result<(u64, u64)> {
+        let src = buff + region.mapping.offset;
+        let start_addr = region.mapping.base_host_virt_addr;
+        let len = region.mapping.size;
+        // Populate whole region from backing mem-file.
+        // This offers an example of how memory can be loaded in RAM,
+        // however this can be adjusted to accommodate use case needs.
+        // SAFETY: this is safe because the parameters are valid.
+        let ret = unsafe {
+            self.copy(src as *const _, start_addr as *mut _, len, true)
+                .map_err(HandlerError::UffdCopy)?
+        };
+
+        // Make sure the UFFD copied some bytes.
+        if ret == 0 {
+            return Err(HandlerError::UffdNoOperation(String::from("copy")));
+        }
+
+        Ok((start_addr, start_addr + ret as u64))
+    }
+
+    fn zero_out(&self, addr: u64) -> Result<(u64, u64)> {
+        let page_size = get_page_size().map_err(HandlerError::PageSize)?;
+
+        // SAFETY: this is safe because the parameters are valid.
+        let ret = unsafe {
+            self.zeropage(addr as *mut _, page_size, true)
+                .map_err(HandlerError::UffdZero)?
+        };
+        // Make sure the UFFD zeroed out some bytes.
+        if ret == 0 {
+            return Err(HandlerError::UffdNoOperation(String::from("zero")));
+        }
+
+        Ok((addr, addr + ret as u64))
+    }
+}
+
+pub struct PageFaultHandler {
+    mem_regions: Vec<MemRegion>,
+    backing_buffer: *const u8,
+    pub uffd: Box<dyn UffdManager>,
```

Comment:
nit: Instead of doing dynamic dispatch via trait objects, we can statically dispatch by parameterizing `PageFultHandler` by `U: UffdManager`, I think.

Reply:
You are right. Done.